### ISSUE TITLE
MoltenVK 1.2.3 Issues

### ISSUE THREAD

**Issue by IsaacMarovitz** [KhronosGroup/MoltenVK#1915]
Hi, Ryujinx recently tried to transition from MVK 1.2.0 to 1.2.3, and we ran into some issues.

The game we ran into issues with was Persona 5 Royale.

MVK 1.2.2
![image](https://github.com/KhronosGroup/MoltenVK/assets/42140194/7d43e392-9269-4cad-a8cf-9f27cae97c08)

MVK 1.2.3 (Under Rosetta)
![image](https://github.com/KhronosGroup/MoltenVK/assets/42140194/284d2fe3-56c5-453e-962d-81b73ca12a90)

MVK 1.2.3 (ARM)
![image](https://github.com/KhronosGroup/MoltenVK/assets/42140194/fe613826-a11c-4c56-b38c-fa6f392c2dd0)

Are there any breaking changes we should be aware of?

**Comment by IsaacMarovitz** [comment]
> Is it possible that this is related to #1838, which was fixed in PR #1913, as part of MoltenVK 1.2.4?

I checked with a master build too, issue still occurs. Worthing noting that Ryujinx was using this extension, I removed it and issue persisted, so I think it's a bug with MVK rather than an issue with the extension itself.

**Comment by IsaacMarovitz** [comment]
Update, if I remove Ryujinx's usage of this extension/explicitly disable it, the issue is resolved. So even if the extension's functionality isn't used, enabling it is enough to start the problem.

**Comment by IsaacMarovitz** [comment]
> Just to clarify...I assume you mean `VK_EXT_external_memory_host`, the subject of the commit you identified?

Yes. Ryujinx already made use of this extension prior to MVK adding support for it; however, even if I disable Ryujinx actually using the extension functionality, ie not changing the memory type, but still let the extension be requested in the Vulkan config, the issue still occurs. If I stop Ryujinx from requesting this extension altogether the issue does not occur.

**Comment by IsaacMarovitz** [comment]
I tested and confirmed, under Rosetta `sysconf(_SC_PAGESIZE)` returns 4096, and under ARM it returns 16384

**Comment by rcaridade145** [comment]
If i remember correctly LibRetro also had some issues with that - https://github.com/libretro/parallel-n64/blob/a03fdcba6b2e9993f050b50112f597ce2f44fa2c/mupen64plus-video-paraLLEl/parallel-rdp/parallel-rdp/rdp_device.cpp#L66 - this link may help

**Comment by billhollings** [comment]
> enabling it is enough to start the problem

This seems odd. Simply enabling the extension in MoltenVK doesn't change anything within MoltenVK itself.

> I tested and confirmed, under Rosetta sysconf(_SC_PAGESIZE) returns 4096, and under ARM it returns 16384

Are you thinking this could be a problem somehow? Is it possible the app is ignoring the larger value in `VkPhysicalDeviceExternalMemoryHostPropertiesEXT::minImportedHostPointerAlignment` under ARM, and always populating it with `4096`? In testing here, even forcing MoltenVK to return `minImportedHostPointerAlignment` as `4096` under ARM doesn't seem to cause a problem with successfully creating buffers on memory with that alignment.

**Comment by IsaacMarovitz** [comment]
> This seems odd. Simply enabling the extension in MoltenVK doesn't change anything within MoltenVK itself.
Seems to be some ongoing debate on that 😅 but that's what I saw in my testing. Could be the extension is being used somewhere else sneakily when its available.

> Are you thinking this could be a problem somehow? 
That was a guestimate based on the fact that it was working under Rosetta but not under ARM and the fact that that is the only code path that gets called when the extension is enabled, even if it's unused. Ryujinx uses C# APIs to get the page size rather than going through MVK, the values should be the same but I can double-check.

```csharp
nint pageAlignedPointer = BitUtils.AlignDown(pointer, Environment.SystemPageSize);
nint pageAlignedEnd = BitUtils.AlignUp((nint)((ulong)pointer + size), Environment.SystemPageSize);
ulong pageAlignedSize = (ulong)(pageAlignedEnd - pageAlignedPointer);
```

**Comment by billhollings** [comment]
> Could be the extension is being used somewhere else sneakily when its available.

Possible, and what I was thinking too. It would be good to try to capture or log activity on that extension, to see it's impact.

> It's worth noting that Ryujinx can work around this problem. But I thought it was worth seeing if it can be resolved

Okay. That's good. Until we can get more info on the interactions and causes, let's put this on the back burner for now.
